sound/soc/sof/ipc4-topology.c

```diff
@@ -263,6 +263,16 @@ static int sof_ipc4_get_audio_fmt(struct snd_soc_component *scomp,
 	return ret;
 }
 
+/* release the memory allocated in sof_ipc4_get_audio_fmt */
+static void sof_ipc4_put_audio_fmt(struct sof_ipc4_available_audio_format *available_fmt)
+
+{
+	kfree(available_fmt->base_config);
+	available_fmt->base_config = NULL;
```

Comment:
if `sof_ipc4_get_audio_fmt()` fails it is freeing up the allocations, but it is not going to NULL available_fmt->base_config.
Perhaps do that and check for NULL before freeing?

On other note:
It looks to me that `sof_ipc4_widget_setup_pcm()` and `sof_ipc4_widget_setup_comp_dai()` also leaks the allocated memory tracked within available_fmt when there is a failure _after_ `sof_ipc4_get_audio_fmt()`?

Reply:
> if `sof_ipc4_get_audio_fmt()` fails it is freeing up the allocations, but it is not going to NULL available_fmt->base_config. Perhaps do that and check for NULL before freeing?

Right, we should set available_fmt->base_config to NULL, otherwise there may be kernel panic. Thanks for pointing it out. We don't need to check NULL before free a point. kfree() will help to check whether it NULL or not.

> 
> On other note: It looks to me that `sof_ipc4_widget_setup_pcm()` and `sof_ipc4_widget_setup_comp_dai()` also leaks the allocated memory tracked within available_fmt when there is a failure _after_ `sof_ipc4_get_audio_fmt()`?

Yes. Will fix it.

Reply:
> > if `sof_ipc4_get_audio_fmt()` fails it is freeing up the allocations, but it is not going to NULL available_fmt->base_config. Perhaps do that and check for NULL before freeing?
> 
> Right, we should set available_fmt->base_config to NULL, otherwise there may be kernel panic. Thanks for pointing it out. We don't need to check NULL before free a point. kfree() will help to check whether it NULL or not.

After second think, the original code is OK. If `sof_ipc4_get_audio_fmt()` fails, the memory will be freed. And ipc_setup() will free `swidget->private` and set `swidget->private = NULL`. So available_fmt members are all freed and will never be used.

Reply:
OK